uvq_pytorch/utils/contentnet.py

```diff
@@ -0,0 +1,144 @@
+import numpy as np
+import pandas as pd
+import torch
+
+from torch import nn, Tensor
+
+from .custom_nn_layers import (
+    Conv2dNormActivationSamePadding,
+    Conv2dSamePadding,
+    Interpolate,
+    MBConvSamePadding,
+)
+from .video_reader import VideoReader
+
+MODEL_PATH = "checkpoint/contentnet_pytorch.pt"
+LABELS_CSV_PATH = "checkpoint/contentnet_labels.csv"
+
+# Output feature size
+DIM_HEIGHT_FEATURE = 16
+DIM_WIDTH_FEATURE = 16
+DIM_CHANNEL_FEATURE = 100
+
+# ContentNet specs
+DIM_LABEL_CONTENT = 3862
+
+
+class ContentNet(nn.Module):
+    def __init__(self, num_classes=DIM_LABEL_CONTENT, dropout=0.2):
+        super().__init__()
+        stochastic_depth_prob_step = 0.0125
+        stochastic_depth_prob = [x * stochastic_depth_prob_step for x in range(16)]
+        self.features = nn.Sequential(
```

Comment:
Can we use the efficient net implementation from torchvision.models? http://docs/document/d/1-DgRdkPvt6LVJesD3bbW6rfbUJiHrgvXy-nIgRKhNpA?resourcekey=0-j_VtCkUxMZtoAYqGSplEEg&tab=t.0#heading=h.8n3y1cyqtsez

Reply:
yes and no. The problem is pytorch (and torchvision's efficinetNet) don't use "same" padding in the convolutions. So the output sizes and as a result the final outputs are a bit different IIRC.
Inititally I had a different implementation that was using the efficientNet implementation but then looping through the architecture and changing all conv2d layers to Conv2dSamePadding custom layer that I have included in this code. I thought this version is cleaner.

Reply:
You are right. We saw this issue before and we used the alternative approach you mentioned (loop + replace conv layer). I think it is fine to explicitly list out all the layers, as long as the model output is similar with TF.

Reply:
The outputs of the two approaches are the same. Let me know if you prefer to change this and I'll prepare the other version.

Reply:
I am OK with the current implementation (perhaps you can add a bit more comment in the code to explain the rationale). We can change that later if needed.

I personally slightly prefer the alternative approach (based on torchvision's efficientnet). The Conv2dSamePadding layer is only needed if you want to copy weights from our TF model. If in the future, you want to retrain everything from scratch, then you probably do not need the Conv2dSamePadding layer.

@yilinwang01 FYI